internal/stores/objects.go

```diff
@@ -259,3 +272,282 @@ func NewJSONObjectStore(dir string) (*JSONObjectStore, error) {
 	}
 	return s, nil
 }
+
+type (
+	SQLObjectStore struct {
+		db *gorm.DB
+	}
+
+	// dbObject describes an object.Object in the database.
+	dbObject struct {
+		// ID uniquely identifies an Object within the database. Since
+		// this ID is also exposed via the API it's a string for
+		// convenience.
+		ID string `gorm:"primaryKey"`
+
+		// Object related fields.
+		Key   string    // json string
+		Slabs []dbSlice `gorm:"constraint:OnDelete:CASCADE;foreignKey:ObjectID;references:ID"` // CASCADE to delete slices too
+	}
+
+	// dbSlice describes a reference to a slab.Slab in the database.
+	dbSlice struct {
+		// ID uniquely identifies a slice in the db.
+		ID uint64 `gorm:"primaryKey"`
+
+		// ObjectID identifies the object the slice belongs to. It is
+		// the foreign key of the object table.
+		ObjectID string `gorm:"index;NOT NULL"`
+
+		// Slice related fields.
+		Slab   dbSlab `gorm:"constraint:OnDelete:CASCADE;foreignKey:ID"` // CASCADE to delete slabs too
+		Offset uint32
+		Length uint32
+	}
+
+	// dbSlab describes a slab.Slab in the database.
+	// NOTE: A Slab is uniquely identified by its key.
+	// NOTE TO REVIEWERS: Is this ok? What if there is no key or we want 2
```

Comment:
definitely ok I would say, we can always adjust it later if need be

Reply:
a slab key should always exist and should always be unique. Encryption overhead is minimal, so no reason not to encrypt. And if you want to change the sharding, you should create a new slab with a new key.